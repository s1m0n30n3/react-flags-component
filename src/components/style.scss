$sizes: 
    "nano" 4px,
    "micro" 6px,
    "mini" 8px,
    "xxxs" 10px,
    "xxs" 12px,
    "xs" 14px,
    "sm" 16px,
    "md" 24px,
    "lg" 32px,
    "xl" 48px,
    "xxl" 80px,
    "xxxl" 80px,
    "jumbo" 96px,
    "maxi" 128px,
    "ultra" 256px;

@each $name, $size in $sizes {
    .fib-size-#{$name} { background-size: cover; width: $size; height: $size; }
}

$flags:
    "ENG" "" "england",
    "WLS" "" "wales",
    "SCT" "" "scotland",
    "AFG" "AF" "afghanistan",
    "ALA" "AX" "aland_islands",
    "ALB" "AL" "albania",
    "DZA" "DZ" "algeria",
    "ASM" "AS" "american_samoa",
    "AND" "AD" "andorra",
    "AGO" "AO" "angola",
    "AIA" "AI" "anguilla",
    "ATG" "AG" "antigua_and_barbuda",
    "ARG" "AR" "argentina",
    "ARM" "AM" "armenia",
    "ABW" "AW" "aruba",
    "AUS" "AU" "australia",
    "AUT" "AT" "austria",
    "AZE" "AZ" "azerbaijan",
    "BHS" "BS" "bahamas",
    "BHR" "BH" "bahrain",
    "BGD" "BD" "bangladesh",
    "BRB" "BB" "barbados",
    "BLR" "BY" "belarus",
    "BEL" "BE" "belgium",
    "BLZ" "BZ" "belize",
    "BEN" "BJ" "benin",
    "BMU" "BM" "bermuda",
    "BTN" "BT" "bhutan",
    "BOL" "BO" "bolivia",
    "BES" "BQ" "bonaire",
    "BIH" "BA" "bosnia_and_herzegovina",
    "BWA" "BW" "botswana",
    "BRA" "BR" "brazil",
    "IOT" "IO" "british_indian_ocean_territory",
    "BRN" "BN" "brunei",
    "BGR" "BG" "bulgaria",
    "BFA" "BF" "burkina_faso",
    "BDI" "BI" "burundi",
    "CPV" "CV" "cape_verde",
    "KHM" "KH" "cambodia",
    "CMR" "CM" "cameroon",
    "CAN" "CA" "canada",
    "CYM" "KY" "cayman_islands",
    "CAF" "CF" "central_african_republic",
    "TCD" "TD" "chad",
    "CHL" "CL" "chile",
    "CHN" "CN" "china",
    "CXR" "CX" "christmas_island",
    "CCK" "CC" "cocos_island",
    "COL" "CO" "colombia",
    "COM" "KM" "comoros",
    "COD" "CD" "democratic_republic_of_congo",
    "COG" "CG" "republic_of_the_congo",
    "COK" "CK" "cook_islands",
    "CRI" "CR" "costa_rica",
    "CIV" "CI" "ivory_coast",
    "HRV" "HR" "croatia",
    "CUB" "CU" "cuba",
    "CUW" "CW" "curacao",
    "CYP" "CY" "cyprus",
    "CZE" "CZ" "czech_republic",
    "DNK" "DK" "denmark",
    "DJI" "DJ" "djibouti",
    "DMA" "DM" "dominica",
    "DOM" "DO" "dominican_republic",
    "ECU" "EC" "ecuador",
    "EGY" "EG" "egypt",
    "SLV" "SV" "el_salvador",
    "GNQ" "GQ" "equatorial_guinea",
    "ERI" "ER" "eritrea",
    "EST" "EE" "estonia",
    "ETH" "ET" "ethiopia",
    "FLK" "FK" "falkland_islands",
    "FRO" "FO" "faroe_islands",
    "FJI" "FJ" "fiji",
    "FIN" "FI" "finland",
    "FRA" "FR" "france",
    "PYF" "PF" "french_polynesia",
    "GAB" "GA" "gabon",
    "GMB" "GM" "gambia",
    "GEO" "GE" "georgia",
    "DEU" "DE" "germany",
    "GHA" "GH" "ghana",
    "GIB" "GI" "gibraltar",
    "GRC" "GR" "greece",
    "GRL" "GL" "greenland",
    "GRD" "GD" "grenada",
    "GUM" "GU" "guam",
    "GTM" "GT" "guatemala",
    "GGY" "GG" "guernsey",
    "GIN" "GN" "guinea",
    "GNB" "GW" "guinea_bissau",
    "GUY" "GY" "guyana",
    "HTI" "HT" "haiti",
    "VAT" "VA" "vatican_city",
    "HND" "HN" "honduras",
    "HKG" "HK" "hong_kong",
    "HUN" "HU" "hungary",
    "ISL" "IS" "iceland",
    "IND" "IN" "india",
    "IDN" "ID" "indonesia",
    "IRN" "IR" "iran",
    "IRQ" "IQ" "iraq",
    "IRL" "IE" "ireland",
    "IMN" "IM" "isle_of_man",
    "ISR" "IL" "israel",
    "ITA" "IT" "italy",
    "JAM" "JM" "jamaica",
    "JPN" "JP" "japan",
    "JEY" "JE" "jersey",
    "JOR" "JO" "jordan",
    "KAZ" "KZ" "kazakhstan",
    "KEN" "KE" "kenya",
    "KIR" "KI" "kiribati",
    "PRK" "KP" "north_korea",
    "KOR" "KR" "south_korea",
    "KWT" "KW" "kuwait",
    "KGZ" "KG" "kyrgyzstan",
    "LAO" "LA" "laos",
    "LVA" "LV" "latvia",
    "LBN" "LB" "lebanon",
    "LSO" "LS" "lesotho",
    "LBR" "LR" "liberia",
    "LBY" "LY" "libya",
    "LIE" "LI" "liechtenstein",
    "LTU" "LT" "lithuania",
    "LUX" "LU" "luxembourg",
    "MAC" "MO" "macao",
    "MKD" "MK" "republic_of_macedonia",
    "MDG" "MG" "madagascar",
    "MWI" "MW" "malawi",
    "MYS" "MY" "malaysia",
    "MDV" "MV" "maldives",
    "MLI" "ML" "mali",
    "MLT" "MT" "malta",
    "MHL" "MH" "marshall_island",
    "MTQ" "MQ" "martinique",
    "MRT" "MR" "mauritania",
    "MUS" "MU" "mauritius",
    "MEX" "MX" "mexico",
    "FSM" "FM" "micronesia",
    "MDA" "MD" "moldova",
    "MCO" "MC" "monaco",
    "MNG" "MN" "mongolia",
    "MNE" "ME" "montenegro",
    "MSR" "MS" "montserrat",
    "MAR" "MA" "morocco",
    "MOZ" "MZ" "mozambique",
    "MMR" "MM" "myanmar",
    "NAM" "NA" "namibia",
    "NRU" "NR" "nauru",
    "NPL" "NP" "nepal",
    "NLD" "NL" "netherlands",
    "NZL" "NZ" "new_zealand",
    "NIC" "NI" "nicaragua",
    "NER" "NE" "niger",
    "NGA" "NG" "nigeria",
    "NIU" "NU" "niue",
    "NFK" "NF" "norfolk_island",
    "MNP" "MP" "northern_marianas_islands",
    "NOR" "NO" "norway",
    "OMN" "OM" "oman",
    "PAK" "PK" "pakistan",
    "PSE" "PS" "palestine",
    "PLW" "PW" "palau",
    "PAN" "PA" "panama",
    "PNG" "PG" "papua_new_guinea",
    "PRY" "PY" "paraguay",
    "PER" "PE" "peru",
    "PHL" "PH" "philippines",
    "PCN" "PN" "pitcairn_islands",
    "POL" "PL" "poland",
    "PRT" "PT" "portugal",
    "PRI" "PR" "puerto_rico",
    "QAT" "QA" "qatar",
    "ROU" "RO" "romania",
    "RUS" "RU" "russia",
    "RWA" "RW" "rwanda",
    "BLM" "BL" "st_barts",
    "KNA" "KN" "saint_kitts_and_nevis",
    "LCA" "LC" "st_lucia",
    "VCT" "VC" "st_vincent_and_the_grenadines",
    "WSM" "WS" "samoa",
    "SMR" "SM" "san_marino",
    "STP" "ST" "sao_tome_and_prince",
    "SAU" "SA" "saudi_arabia",
    "SEN" "SN" "senegal",
    "SRS" "RS" "serbia",
    "SYC" "SC" "seychelles",
    "SLE" "SL" "sierra_leone",
    "SGP" "SG" "singapore",
    "SXM" "SX" "sint_maarten",
    "SVK" "SK" "slovakia",
    "SVN" "SI" "slovenia",
    "SLB" "SB" "solomon_islands",
    "SOM" "SO" "somalia",
    "ZAF" "ZA" "south_africa",
    "SSD" "SS" "south_sudan",
    "ESP" "ES" "spain",
    "LKA" "LK" "sri_lanka",
    "SDN" "SD" "sudan",
    "SUR" "SR" "suriname",
    "SWE" "SE" "sweden",
    "CHE" "CH" "switzerland",
    "SYR" "SY" "syria",
    "TWN" "TW" "taiwan",
    "TJK" "TJ" "tajikistan",
    "TZA" "TZ" "tanzania",
    "THA" "TH" "thailand",
    "TLS" "TL" "east_timor",
    "TGO" "TG" "togo",
    "TKL" "TK" "tokelau",
    "TON" "TO" "tonga",
    "TTO" "TT" "trinidad_and_tobago",
    "TUN" "TN" "tunisia",
    "TUR" "TR" "turkey",
    "TKM" "TM" "turkmenistan",
    "TCA" "TC" "turks_and_caicos",
    "TUV" "TV" "tuvalu",
    "UGA" "UG" "uganda",
    "UKR" "UA" "ukraine",
    "ARE" "AE" "united_arab_emirates",
    "GBR" "GB" "united_kingdom",
    "USA" "US" "united_states",
    "URY" "UY" "uruguay",
    "UZB" "UZ" "uzbekistan",
    "VUT" "VU" "vanuatu",
    "VEN" "VE" "venezuela",
    "VNM" "VN" "vietnam",
    "VGB" "VG" "british_virgin_islands",
    "VIR" "VI" "virgin_islands",
    "YEM" "YE" "yemen",
    "ZMB" "ZM" "zambia",
    "ZWE" "ZW" "zimbabwe";

@each $alpha3, $alpha2, $name in $flags {
    .fib-rectangular-#{$alpha3},
    .fib-rectangular-#{$alpha2} {
        background-image: url('../assets/rectangular/#{$name}.svg');
    }
    .fib-rounded-#{$alpha3},
    .fib-rounded-#{$alpha2} {
        background-image: url('../assets/rounded/#{$name}.svg');
    }
}
