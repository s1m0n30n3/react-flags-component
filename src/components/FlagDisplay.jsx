import React from "react";
import PropTypes from "prop-types";

import "./style.scss";

export const FlagDisplay = ({
  countryCode = "AFG",
  size = "md",
  shape = "rectangular",
  as: Tag = "div",
  customHeight,
  customWidth,
  ...props
}) => {
  const customHeightProp = customHeight ? { height: customHeight } : {};
  const customWidthProp = customWidth ? { width: customWidth } : {};

  const customSizeProps =
    customHeight || customHeight
      ? { style: { ...customHeightProp, ...customWidthProp } }
      : {};

  return (
    <Tag
      className={`fib-${shape}-${countryCode} fib-size-${size}`}
      {...customSizeProps}
      {...props}
    />
  );
};

FlagDisplay.propTypes = {
  countryCode: PropTypes.string,
  size: PropTypes.string,
  shape: PropTypes.string,
  as: PropTypes.string,
  customHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  customWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
};
