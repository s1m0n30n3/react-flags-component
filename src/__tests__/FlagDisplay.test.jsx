import React from "react";
import renderer from "react-test-renderer";
import { describe, expect, test } from "vitest";

import { FlagDisplay } from "../components/FlagDisplay";

describe("FlagDisplay", () => {
  test("renders correctly", () => {
    const component = renderer.create(<FlagDisplay />);

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("renders correctly with default css classes", () => {
    const component = renderer.create(<FlagDisplay />);

    const tree = component.toJSON();
    const className = tree.props.className;

    expect(className).toBe("fib-rectangular-AFG fib-size-md");
  });

  test("renders correctly with Italian flag (ITA - alpha 3) and with size medium (md) in a div", () => {
    const component = renderer.create(
      <FlagDisplay countryCode="ITA" size="md" />
    );

    const tree = component.toJSON();
    const {
      type,
      props: { className },
    } = tree;

    expect(className).toBe("fib-rectangular-ITA fib-size-md");
    expect(type).toBe("div");
  });

  test("renders correctly with Italian flag (ITA - alpha 3) and with size medium (md) in a div", () => {
    const component = renderer.create(
      <FlagDisplay countryCode="ITA" size="md" />
    );

    const tree = component.toJSON();
    const {
      type,
      props: { className },
    } = tree;

    expect(className).toBe("fib-rectangular-ITA fib-size-md");
    expect(type).toBe("div");
  });

  test("renders correctly with Italian flag (IT - alpha 2) and with size ultra (ultra) in a hyperlink tag (a)", () => {
    const component = renderer.create(
      <FlagDisplay as="a" countryCode="IT" size="ultra" />
    );

    const tree = component.toJSON();
    const {
      type,
      props: { className },
    } = tree;

    expect(className).toBe("fib-rectangular-IT fib-size-ultra");
    expect(type).toBe("a");
  });

  test("renders correctly with customWidth and customHeight in a div", () => {
    const component = renderer.create(
      <FlagDisplay as="a" customHeight={10} customWidth={10} countryCode="IT" />
    );

    const tree = component.toJSON();
    const {
      type,
      props: {
        className,
        style: { height, width },
      },
    } = tree;

    expect(className).toBe("fib-rectangular-IT fib-size-md");
    expect(type).toBe("a");
    expect(height).toBe(10);
    expect(width).toBe(10);
  });
});
