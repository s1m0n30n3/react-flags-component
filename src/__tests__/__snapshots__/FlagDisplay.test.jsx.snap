// Vitest Snapshot v1

exports[`FlagDisplay > renders correctly 1`] = `
<div
  className="fib-rectangular-AFG fib-size-md"
/>
`;
